File: mac/DeskLayer/DeskLayerTests/RecordingCanvas.swift
//
//  RecordingCanvas.swift
//  DeskLayerTests
//
//  A CanvasJSExports implementation that records every ctx call as data
//  instead of drawing. The op log, serialized canonically, is the artifact
//  under shared/conformance/canvas/golden: any port's ctx bridge must
//  produce the same log for the same fixture. The exact recording rules
//  live in shared/conformance/runner-notes.md — change them there first.
//
//  measureText returns width = 7 × UTF-16 length, a deterministic stub so
//  goldens don't depend on any platform's text stack.
//

import Foundation
import JavaScriptCore
@testable import DeskLayer

final class RecordingCanvas: NSObject, CanvasJSExports, @unchecked Sendable {
    private(set) var ops: [Any] = []
    private let widthPts: Double
    private let heightPts: Double

    /// Same bridge production uses: PropertyValue.jsValue or nil.
    var propertyProvider: ((String) -> Any?)?

    init(width: Double, height: Double) {
        widthPts = width
        heightPts = height
        super.init()
    }

    /// Called by the conformance runner before each render() call.
    func mark(frame index: Int) {
        ops.append(["op": "frame", "index": Double(index)])
    }

    private func record(_ op: String, _ args: [Any]) {
        ops.append(["op": op, "args": args])
    }

    private func recordSet(_ name: String, _ value: Any) {
        ops.append(["op": "set", "name": name, "value": value])
    }

    // MARK: - Properties

    var fillStyle: String = "#000000" { didSet { recordSet("fillStyle", fillStyle) } }
    var strokeStyle: String = "#000000" { didSet { recordSet("strokeStyle", strokeStyle) } }
    var lineWidth: Double = 1 { didSet { recordSet("lineWidth", lineWidth) } }
    var lineCap: String = "butt" { didSet { recordSet("lineCap", lineCap) } }
    var lineJoin: String = "miter" { didSet { recordSet("lineJoin", lineJoin) } }
    var globalAlpha: Double = 1 { didSet { recordSet("globalAlpha", globalAlpha) } }
    var font: String = "13px Helvetica" { didSet { recordSet("font", font) } }
    var width: Double { widthPts }
    var height: Double { heightPts }

    // MARK: - State & transforms

    func save() { record("save", []) }
    func restore() { record("restore", []) }
    func translate(_ x: Double, _ y: Double) { record("translate", [x, y]) }
    func rotate(_ angle: Double) { record("rotate", [angle]) }
    func scale(_ x: Double, _ y: Double) { record("scale", [x, y]) }

    // MARK: - Rects

    func clearRect(_ x: Double, _ y: Double, _ w: Double, _ h: Double) { record("clearRect", [x, y, w, h]) }
    func fillRect(_ x: Double, _ y: Double, _ w: Double, _ h: Double) { record("fillRect", [x, y, w, h]) }
    func strokeRect(_ x: Double, _ y: Double, _ w: Double, _ h: Double) { record("strokeRect", [x, y, w, h]) }

    // MARK: - Paths

    func beginPath() { record("beginPath", []) }
    func closePath() { record("closePath", []) }
    func moveTo(_ x: Double, _ y: Double) { record("moveTo", [x, y]) }
    func lineTo(_ x: Double, _ y: Double) { record("lineTo", [x, y]) }

    func arc(_ x: Double, _ y: Double, _ r: Double, _ startAngle: Double, _ endAngle: Double, _ anticlockwise: Bool) {
        record("arc", [x, y, r, startAngle, endAngle, anticlockwise])
    }

    func fill() { record("fill", []) }
    func stroke() { record("stroke", []) }

    // MARK: - Text

    func fillText(_ text: String, _ x: Double, _ y: Double) { record("fillText", [text, x, y]) }

    func measureText(_ text: String) -> [String: Double] {
        record("measureText", [text])
        return ["width": Double(text.utf16.count) * 7]
    }

    // MARK: - Images

    func drawImage(_ name: String, _ x: Double, _ y: Double, _ w: Double, _ h: Double) {
        record("drawImage", [name, x, y, w, h])
    }

    // MARK: - Properties bridge

    func getProp(_ name: String) -> Any? {
        record("getProp", [name])
        return propertyProvider?(name)
    }
}

File: win/src/DeskLayer.Core/Conformance/RecordingCanvas.cs
// Records every ctx call as data instead of drawing — the C# twin of the mac
// RecordingCanvas. The recording rules live in
// shared/conformance/runner-notes.md; both implementations must match them
// exactly, or the golden diff is the bug report.
//
// Members are lowercase to match the JS contract (Jint binds case-sensitively).

namespace DeskLayer.Core.Conformance;

public sealed class RecordingCanvas
{
    public List<object> Ops { get; } = new();
    private readonly double widthPts;
    private readonly double heightPts;

    /// Same bridge production uses: PropertyValue.BridgeValue or null.
    public Func<string, object?>? PropertyProvider { get; set; }

    public RecordingCanvas(double width, double height)
    {
        widthPts = width;
        heightPts = height;
    }

    public void MarkFrame(int index) =>
        Ops.Add(new Dictionary<string, object> { ["op"] = "frame", ["index"] = (double)index });

    private void Record(string op, params object[] args) =>
        Ops.Add(new Dictionary<string, object> { ["op"] = op, ["args"] = args.ToList() });

    private void RecordSet(string name, object value) =>
        Ops.Add(new Dictionary<string, object> { ["op"] = "set", ["name"] = name, ["value"] = value });

    // ---- properties (recorded on every assignment) ----

    private string fillStyleValue = "#000000";
    public string fillStyle { get => fillStyleValue; set { fillStyleValue = value; RecordSet("fillStyle", value); } }

    private string strokeStyleValue = "#000000";
    public string strokeStyle { get => strokeStyleValue; set { strokeStyleValue = value; RecordSet("strokeStyle", value); } }

    private double lineWidthValue = 1;
    public double lineWidth { get => lineWidthValue; set { lineWidthValue = value; RecordSet("lineWidth", value); } }

    private string lineCapValue = "butt";
    public string lineCap { get => lineCapValue; set { lineCapValue = value; RecordSet("lineCap", value); } }

    private string lineJoinValue = "miter";
    public string lineJoin { get => lineJoinValue; set { lineJoinValue = value; RecordSet("lineJoin", value); } }

    private double globalAlphaValue = 1;
    public double globalAlpha { get => globalAlphaValue; set { globalAlphaValue = value; RecordSet("globalAlpha", value); } }

    private string fontValue = "13px Helvetica";
    public string font { get => fontValue; set { fontValue = value; RecordSet("font", value); } }

    public double width => widthPts;
    public double height => heightPts;

    // ---- state & transforms ----

    public void save() => Record("save");
    public void restore() => Record("restore");
    public void translate(double x, double y) => Record("translate", x, y);
    public void rotate(double angle) => Record("rotate", angle);
    public void scale(double x, double y) => Record("scale", x, y);

    // ---- rects ----

    public void clearRect(double x, double y, double w, double h) => Record("clearRect", x, y, w, h);
    public void fillRect(double x, double y, double w, double h) => Record("fillRect", x, y, w, h);
    public void strokeRect(double x, double y, double w, double h) => Record("strokeRect", x, y, w, h);

    // ---- paths ----

    public void beginPath() => Record("beginPath");
    public void closePath() => Record("closePath");
    public void moveTo(double x, double y) => Record("moveTo", x, y);
    public void lineTo(double x, double y) => Record("lineTo", x, y);
    public void arc(double x, double y, double r, double startAngle, double endAngle, bool anticlockwise) =>
        Record("arc", x, y, r, startAngle, endAngle, anticlockwise);
    public void fill() => Record("fill");
    public void stroke() => Record("stroke");

    // ---- text ----

    public void fillText(string text, double x, double y) => Record("fillText", text, x, y);

    public MeasureResult measureText(string text)
    {
        Record("measureText", text);
        // Deterministic stub: 7 × UTF-16 code-unit count (runner-notes.md).
        return new MeasureResult(text.Length * 7.0);
    }

    // ---- images / properties ----

    public void drawImage(string name, double x, double y, double w, double h) =>
        Record("drawImage", name, x, y, w, h);

    public object? getProp(string name)
    {
        Record("getProp", name);
        return PropertyProvider?.Invoke(name);
    }
}

public sealed class MeasureResult
{
    public double width { get; }
    public MeasureResult(double width) => this.width = width;
}
